Doc 11873 3.2 GA (#893)
couchbaselabs/docs-couchbase-lite: modules/android/pages/vector-search.adoc

When a model considers the features of a given input as similar, the distance between the vector embeddings will be short.
Vector embeddings are stored within embedded vector indexes.

The current supported formats for Vector embeddings are:

* An array of 32-bit floats.

* A Base64 string that encodes the binary representation of an array of 32-bit float numbers in https://en.wikipedia.org/wiki/Endianness[Little-endian] format.

=== About Vector Indexes

Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 

> Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.

Reviewer: Passive voice here makes it unclear as to who actually odes the configuration. Can you rephrase to make this clearer and simpler? Is this something I need to do? If not, how does it happen?

Author: It's overly wordy I admit as I copied verbatim. Did some research into the object and how it's stored. Likes like you can encode floats into a string format to store or transmit them in text-based formats, like JSON.

Author: Simplified

> If training is triggered during a query, the query may take longer to return results.

Reviewer: Be aware that vector index training can affect query performance.

I'd move the note about the required number of vectors up, directly after you introduce the concept of training.

Author: Done

If a query is executed against the index before it is trained, a full scan of the vectors will be performed. 

NOTE: If there are insufficient vectors in the database for training, a warning message will be logged, indicating the required number of vectors.

=== About Lazy Vector Indexes

> You can use Lazy Vector Indexes(lazy index) to use a ML model that is not available locally on the device and to create vector indexes without having vector embeddings in the documents.

Reviewer: I don't think you need to i-cap lazy vector index....this is a standard industry term, not a branded feature.

also, spell out machine learning (ML)

Author: Done

> Lazy indexing is an asynchronous process that provides developers full control during the following:

Reviewer 2: Not sure this reads correctly. I think 'during the following` doesn't match up with the list below.; it's not explaining during what.

Author: Edited from previous feedback

> Lazy indexing is an asynchronous process that provides developers full control during the following:

Reviewer: Lazy indexing is an asynchronous process that provides developers full control of:
* When to update the index
* The number of vectors to update to the index
* Whether to cancel or skip certain indexes when the ML model is unavailable or has failed.

Author: Much better, thanks! Changed.

* When to update the index.

* The number of vectors to update to the index.

* Cancelling or skipping select indexes when the model is unavailable or has failed.

NOTE: Updating in lazy index is an independent process from saving document operations.

> You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.

Reviewer 2: ```suggestion
You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process,  and retrying the index update process at a later point when the ML model is available.
```

Author: Content removed and condensed down

> You can use lazy indexing for situations when the ML model may be unreachable or not functioning, scheduling the index updating process on demand and handling situations during failed update attempts by skipping the process and retrying the index update process at a later point when the ML model is available.

Reviewer: I don't really understand this sentence....are you saying:

Use lazy indexing to handle situations when the machine learning model is unreachable or not functioning correctly.  In these cases, you can choose to schedule an on-demand index update later, when the ML model is available.

Author: Made text more concise. Should be clearer now.

> IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 

Reviewer: Suggest you be more explicit about what proces you mean here. Do you mean Updating a lazy index?  Also consider combining this info with the previous note

Author: Actioned

* An array of 32-bit floats.

* A Base64 string that encodes the binary representation of an array of 32-bit float numbers in https://en.wikipedia.org/wiki/Endianness[Little-endian] format.

=== About Vector Indexes

Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 
Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.
If training is triggered during a query, the query may take longer to return results.
If a query is executed against the index before it is trained, a full scan of the vectors will be performed. 

NOTE: If there are insufficient vectors in the database for training, a warning message will be logged, indicating the required number of vectors.

=== About Lazy Vector Indexes

IMPORTANT: Lazy index is not an automatic process, you will need to manually schedule the index updates. 

> Lazy Vector Indexes(lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:

Reviewer 3: spacing

```suggestion
Lazy Vector Indexes (lazy index) is Couchbase Lite specific functionality that updates indexes asynchronously, satisfying the following use cases:
```

Author: Done

> * Documents have been added to the application by the end-user with no available Machine Learning(ML) model to generate the vectors. - You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 

Reviewer 3: Minor spacing and punctuation edits

```suggestion
* Documents have been added to the application by the end-user with no available Machine Learning (ML) model to generate the vectors.
You can use lazy indexing to schedule updating the index of said documents when a ML model is available. 
```

Author: Done

> * The remote ML model used stops working or has intermittent availability, causing a failed update - With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.

Reviewer 3: Minor punctuation edits

```suggestion
* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.
```

Author: Done

You can use lazy indexing to schedule updates to the index of said documents when a ML model is available. 

* The remote ML model used stops working or has intermittent availability, causing a failed update.
With lazy index, you can skip documents that fail to update and schedule the index process to be retried at a later date.

Lazy indexing is an asynchronous process that provides developers full control of:

* When to update the index.

* The number of vectors to update to the index.

* Whether to cancel or skip certain indexes when the model is unavailable or has failed.

NOTE: Updating in lazy index is an independent process from saving document operations.

See here for examples of how to xref:android:working-with-vector-search.adoc#create-a-lazy-vector-index[use lazy index in your applications].

Lazy indexing is an alternate approach to using the standard predictive model with regular vector indexes which handle the indexing process automatically. 
The table below compares the two processes.

.Regular Vector Indexes vs Lazy Vector Indexes
[cols ="3*"]
|===
|Feature |Regular Index |Lazy Index

|Update when documents are changed
|image:ROOT:yes.png[]

> |image:ROOT:no.png[]

Reviewer 4: Having spent some "quality time" recently with tables... please put a new-line in between rows, for sanity's sake 😂

The general guideline for the amount of Centroids is approximately the square root of the number of documents.

> === About Probes

Reviewer: This is a little jarring--I don't know why you re mentioning probes....start with a statement to let user know what these are

Author: This is now addressed with clearer explanation surrounding probes.

Only the direction of the vectors matter.
The smaller the distance value, the more similar the vectors are.

== Hybrid Vector Search

> You can use Vector Search with regular SQL++ query language. 

Reviewer 2: ```suggestion
You can use Vector Search with the regular SQL++ query language. 
```

Author: Changed for a better explanation of the topic.

Vector search will be performed on the documents that have been filtered based on the criteria specified in the WHERE clause.

> No LIMIT clause is required for Hybrid Vector Search.

Reviewer 2: ```suggestion
The LIMIT clause is not required for Hybrid Vector Search.
```

See the https://www.couchbase.com/blog/hybrid-search/[Hybrid Search blog post] for more information about Hybrid Search.

> IMPORTANT: You cannot `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.

Reviewer 2: ```suggestion
IMPORTANT: You cannot  use `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.
```

Author: Done, moved to working with page

> IMPORTANT: You cannot `OR` with Hybrid Vector Search or when using `APPROX_VECTOR_DISTANCE()` with other expressions in the `WHERE` clause.

Reviewer: I don't understand what you mean in the "important" note.

Author: Actioned, moved to working with page near relevant snippets for better context

== Next Steps

> Now that you're familiar with the core concepts of vector search, it's time to put your knowledge into practice and start building AI applications.

Reviewer: I would strike the Next Steps section entirely. I don't think it adds value.

Author: Done
